### ISSUE TITLE
Nodeset Compiler: error C2099: initializer is not a constant

### ISSUE THREAD

**Issue by Pro** [open62541/open62541#4687]
## Description

With version 1.2.2 the nodeset compiler was changed that it is adding the custom types array.
Specifically the commit https://github.com/open62541/open62541/commit/c706a2d8dc0f0e53109123bafba9634f4398a24d of MR #4015 

This fails on Windows when using a shared build with:

`error C2099: initializer is not a constant`

We (@Mo-Tay) found the issue by adding open62541 1.2.2 to the conan center:

https://github.com/conan-io/conan-center-index/pull/7314

There it is using a sample XML file (https://github.com/conan-io/conan-center-index/blob/22ffee62151d68deb590082e906aa3b8ca6ab8f9/recipes/open62541/all/test_package/FooFlt.NodeSet2.xml) to test the nodeset compiler.

The generated files contain this:

`types_foo_flt_generated.h`

```cpp
// ...

/**
 * Every type is assigned an index in an array containing the type descriptions.
 * These descriptions are used during type handling (copying, deletion,
 * binary encoding, ...). */
#define UA_TYPES_FOO_FLT_COUNT 1
extern UA_EXPORT const UA_DataType UA_TYPES_FOO_FLT[UA_TYPES_FOO_FLT_COUNT];

// ...
```

And then (due to previously mentioned commit):

`namespace_foo_flt_generated.c`:

```cpp
static UA_DataTypeArray customUA_TYPES_FOO_FLT = {
    NULL,
    UA_TYPES_FOO_FLT_COUNT,
    UA_TYPES_FOO_FLT
};

UA_StatusCode namespace_foo_flt_generated(UA_Server *server) {
UA_StatusCode retVal = UA_STATUSCODE_GOOD;
/* Use namespace ids generated by the server */
UA_UInt16 ns[2];
ns[0] = UA_Server_addNamespace(server, "http://opcfoundation.org/UA/");
ns[1] = UA_Server_addNamespace(server, "https://new.foo.com/zebra-compression/flattening-and-subspacefolding/UA/");

/* Load custom datatype definitions into the server */
if(UA_TYPES_FOO_FLT_COUNT > 0) {
customUA_TYPES_FOO_FLT.next = UA_Server_getConfig(server)->customDataTypes;
UA_Server_getConfig(server)->customDataTypes = &customUA_TYPES_FOO_FLT;

}
```

Especially note the 4th line which is using `UA_TYPES_FOO_FLT` for the initialization of the variable, and this is where the error comes from.

See also https://stackoverflow.com/a/3025106

## Background Information / Reproduction Steps

Probably only reproducible on windows?

1. Clone the branch (or when merged the master) from https://github.com/conan-io/conan-center-index/pull/7314
2. `cd conan-center-index/recipes/open62541/all`
3. `conan create . "1.2.2@" -o open62541:shared=True`

That should fail with the error mentioned above.

**Note that this is only relevant if you build it as a C application, and not a CXX.**

Used CMake options:

```bash
cmake -DUA_NAMESPACE_ZERO=<YOUR_OPTION> <ANY_OTHER_OPTIONS> ..
```

## Checklist

Please provide the following information:

 - [ ] open62541 Version (release number or git tag): 1.2.2
 - [ ] Other OPC UA SDKs used (client or server): 
 - [ ] Operating system: Windows
 - [ ] Logs (with `UA_LOGLEVEL` set as low as necessary) attached
 - [ ] Wireshark network dump attached
 - [ ] Self-contained code example attached
 - [ ] Critical issue

**Comment by Kischy** [comment]
Is anyone working on this? I would like to use a conan package of open62541 that allows DLLs on Windows.

**Comment by thomasleung2018** [comment]
I also have this issue. I can't use the generated file in Windows. But it works in Linux.

**Comment by jpfr** [comment]
The problem appears to be the linking against a dll.
The linker complains about the memory address becoming available too late.

It works when compiling a static executable.

**Comment by jpfr** [comment]
This gets fixed in #6990.
